helixbass/tsc-rust
# Store a reference to Rc<Node> wrapper in AST nodes

The starting point here is basically option (c) from [this comment](https://github.com/helixbass/tsc-rust/pull/5#discussion_r771830998)

In this PR:
- add a `_node_wrapper: RefCell<Option<Weak<Node>>>` field to the shared `BaseNode` struct for holding a weak reference to the "wrapper" `Rc<Node>` to which it belongs, and add `node_wrapper()`/`set_node_wrapper()` to `NodeInterface` (so that effectively you can get a reference to your wrapping `Rc<Node>` from any AST node type/sub-variant)
- to wire up setting that node wrapper reference upon node creation, make the `#[ast_node]` macro generate an `impl From<[node type]> for Rc<Node>` (along with other existing boilerplate "wrapping" `impl From`'s) that encapsulates setting that node wrapper right after it's created (and then always use `.into()` for creating `Rc<Node>`'s during parsing)
- with that in place, stop passing around an additional `wrapper` argument when access to the `Rc<Node>` wrapper is needed, and make a lot of functions not take `Rc<Node>`'s that don't really care about its "`Rc`-ness" (since we can now obtain an `Rc<Node>` wrapper reference from anything implementing `NodeInterface`)

To test:
Things should still behave the same as of #7 (if you do the testing from #9 it'd basically do the same thing as described there but would panic in a slightly different way because of the in-progress work from `variable-declaration-type-checking` (see below))

Based on `variable-declaration-type-checking` (a branch I've been working on but haven't gotten up for review yet)
